File: src/main/java/shopping/product/service/ProductService.java
package shopping.product.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import shopping.product.domain.Product;
import shopping.product.domain.ProductRepository;

import java.util.NoSuchElementException;

@Service
public class ProductService {

    private final ProductRepository productRepository;
    private final ProfanityValidator profanityValidator;

    public ProductService(ProductRepository productRepository, ProfanityValidator profanityValidator) {
        this.productRepository = productRepository;
        this.profanityValidator = profanityValidator;
    }

    public Product save(Product product) {
        validateProductName(product);

        return productRepository.save(product);
    }

    public Product findProductById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("상품을 찾을 수 없습니다."));
    }

    public Page<Product> findProducts(Pageable pageable) {
        return productRepository.findAll(pageable);
    }

    @Transactional
    public Product update(Long id, Product product) {
        validateProductName(product);
        Product found = productRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("상품을 찾을 수 없습니다."));
        found.update(product);
        return productRepository.save(found);
    }

    public void deleteById(Long id) {
        productRepository.deleteById(id);
    }

    private void validateProductName(Product product) {
        if (profanityValidator.containsProfanity(product.getName())) {
            throw new IllegalArgumentException("비속어가 포함되어 있습니다.");
        }
    }
}

File: src/main/kotlin/shopping/application/ProductService.kt
package shopping.application

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import shopping.application.dto.CreateProductCommand
import shopping.application.dto.ProductResult
import shopping.application.dto.UpdateProductCommand
import shopping.domain.Product
import shopping.domain.ProductRepository
import shopping.domain.ProfanityFilter
import shopping.support.error.CoreException
import shopping.support.error.ErrorType

@Service
@Transactional(readOnly = true)
class ProductService(
    private val productRepository: ProductRepository,
    private val profanityFilter: ProfanityFilter,
) {
    @Transactional
    fun create(command: CreateProductCommand): Long {
        validateProfanity(command.name)

        val product =
            Product(
                name = command.name,
                price = command.price,
                imageUrl = command.imageUrl,
            )
        return productRepository.save(product).id
    }

    @Transactional
    fun update(
        id: Long,
        command: UpdateProductCommand,
    ) {
        validateProfanity(command.name)

        val product = findProductById(id)
        product.update(
            name = command.name,
            price = command.price,
            imageUrl = command.imageUrl,
        )
        productRepository.save(product)
    }

    fun getProduct(id: Long): ProductResult {
        val product = findProductById(id)
        return ProductResult.from(product)
    }

    fun getProducts(): List<ProductResult> {
        val products = productRepository.findAll()
        return products.map { ProductResult.from(it) }
    }

    @Transactional
    fun delete(id: Long) {
        val product = findProductById(id)
        productRepository.deleteById(product.id)
    }

    private fun validateProfanity(name: String) {
        if (profanityFilter.containsProfanity(name)) {
            throw CoreException(ErrorType.PRODUCT_PROFANITY_DETECTED)
        }
    }

    private fun findProductById(id: Long): Product = productRepository.findById(id) ?: throw CoreException(ErrorType.PRODUCT_NOT_FOUND)
}
